### ISSUE TITLE
Investigate options for continuous integration

### ISSUE THREAD

**Issue by mwpenny** [mwpenny/portal64-still-alive#63]
It would be easier to avoid breaking the build by implementing CI builds. For example, using GitHub Actions to compile whenever a branch is updated. The complicating factor here is that the Portal game files and compiled Portal 64 ROM must be kept private (see [N64 Libraries](https://github.com/mwpenny/portal64-still-alive/blob/master/documentation/n64_library_usage.md)).

[SteamCMD](https://developer.valvesoftware.com/wiki/SteamCMD) can be used to download game files and credentials/API keys can be stored using GitHub Actions [secrets](https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions). This could be slow and so [caching](https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows) is worth looking into. Based on GitHub Actions documentation, files generated during jobs are not exposed by default but this should be confirmed, particularly if using caching.

As for keeping the compiled ROM private, according to [Storing workflow data as artifacts](https://docs.github.com/en/actions/using-workflows/storing-workflow-data-as-artifacts), again it appears that exposing build output is opt-in. Confirm this as well.

**Important:** Sensitive data must not be accessible through specially-crafted pull requests. See https://securitylab.github.com/research/github-actions-preventing-pwn-requests/. One way to avoid this is to require approval for workflow runs or to only trigger on events such as pull request approval.

If GitHub-hosted runners can't provide the necessary data security, a few other options are:
* [Self-hosted runner](https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners/about-self-hosted-runners) (I (@mwpenny) can host this if it is needed)
* Supporting compilation (but not running) with missing/empty game data

In summary:
- Determine what data (if any) will be exposed if using GitHub actions with GitHub-hosted runners
- If GitHub Actions with GitHub-hosted runners **are** viable, describe what is needed to implement CI builds using this approach
- If GitHub Actions with GitHub-hosted runners **are not** viable, describe alternative approach(es)

**Comment by mwpenny** [comment]
Thanks for prototyping this! It looks very promising. 

> At least I'm missing something, workflow artifacts aren't exposed by default and Steam credentials are stored as secret, so there shouldn't be any leaks of sensitive data.

Yes, based on the documentation and your run you linked to, this is confirmed :)

> There's also some room for improvement

I'm happy to see it's so quick, and agree caching isn't worth it in the first pass given the game was downloaded in ~1m. Getting to it in the future would be a nice (albeit small) improvement.

Similarly, installing dependencies at the same time the game data is downloaded will save ~1m based on the numbers in the run you shared (assuming no caching). I did some more reading and there doesn't seem to be a nice way to run steps in parallel. We'd have to do some scripting to accomplish this, which wouldn't be difficult.

So we have two possible ways to save that ~1m. My leaning is the former, since it results in the workflow doing less work. But as mentioned it's not critical.

> But the concern about PR vulnerability seems serious. [...] Perhaps a possible solution would be to translate the Makefile into a Github Actions workflow, which in theory cannot be tampered with by a malicious PR, but this does not exclude the possibility of exposing a vulnerability at some step of the workflow.

Yes, it's tricky. If re-creating the Makefile logic, the workflow would no longer be running the exact same steps a user would, meaning bugs could slip through or the two build processes could diverge over time. Regardless, there would still be the possibility of malicious code somewhere else we don't handle (e.g., in the various scripts called during the build).

I agree that CI is useful even if only after PR. But ideally contributors can be alerted to build issues before the merge, without having to configure their own CI. The less friction the better. I think it's good enough to just require approval to trigger workflow runs from forks. For example, using [Environments](https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment).

**Comment by GiacomoGarbin** [comment]
> I did some more reading and there doesn't seem to be a nice way to run steps in parallel. We'd have to do some scripting to accomplish this, which wouldn't be difficult.

Sorry, you are right, steps are executed sequentially, I was thinking about jobs, which are executed in parallel, but I forgot that [inter-job communication](https://docs.github.com/en/actions/using-workflows/storing-workflow-data-as-artifacts#passing-data-between-jobs-in-a-workflow) is done via upload/download artifacts, which is what we want to avoid. But yes, parallelization should also be possible via scripting.

> If re-creating the Makefile logic, the workflow would no longer be running the exact same steps a user would, meaning bugs could slip through or the two build processes could diverge over time. Regardless, there would still be the possibility of malicious code somewhere else we don't handle (e.g., in the various scripts called during the build).

Yes, the workflow approach is a bad idea from multiple points of view.

> So we have two possible ways to save that ~1m. My leaning is the former, since it results in the workflow doing less work. But as mentioned it's not critical.

I haven't read up on caching yet, but it's definitely worth having a look.

> I agree that CI is useful even if only after PR.

Once we have the CI up and running, I think it would be useful to test the different setups ("recommended", docker and manual) on Ubuntu, assuming we want to maintain them all, and then maybe test them on Windows and macOS too, since the runners hosted on GitHub offer this opportunity. And all these tests can be performed in parallel jobs.

> I think it's good enough to just require approval to trigger workflow runs from forks. For example, using [Environments](https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment).

[Environment secrets](https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment#environment-secrets):
> Secrets stored in an environment are only available to workflow jobs that reference the environment. If the environment requires approval, a job cannot access environment secrets until one of the required reviewers approves it.

This sounds perfect.

**Comment by mwpenny** [comment]
> Once we have the CI up and running, I think it would be useful to test the different setups

Yes, agreed. That will help detect problems that only affect specific build setups.

Right now Windows and Mac won't work as-is with the manual setup steps, and Docker should (in theory) be the same on all platforms. Let's start with native (`setup.sh`) + Docker on Ubuntu. The manual steps are the same as what is in `setup.sh`.
